deep_qa/training/trainer.py

```diff
@@ -281,9 +285,13 @@ def train(self):
         train_data, val_data, test_data = self.prepare_data(self.train_files, self.max_training_instances,
                                                             self.validation_files,
                                                             self.test_files)
-        self.training_dataset, self.train_input, self.train_labels = train_data
-        self.validation_dataset, self.validation_input, self.validation_labels = val_data
-        self.test_dataset, self.test_input, self.test_labels = test_data
+        if not self.use_data_generator:
```

Comment:
randomly decided to take a look at this (i hope you don't mind...), but I don't quite see the benefit of a data generator here. You still have to compute the `train_data, val_data, test_data` from `self.prepare_data`, right? So I don't quite see how this is any different than just popping things off of a list, i.e. I think that you'd still have to precompute `X` and `y` with this approach, which makes me question the value of a generator.

I've been using a modified version of the dataset code in this repo to structure some personal work, and I (think, haven't actually tested it yet) managed to get a generator working by replacing the equivalent of `self._prepare_data` with a function that returns a generator, vs lists. This involved making the `pad_instances` function in Datasets return a generator, so you essentially pad "on-demand", but I haven't figured out how to get it working here...